include/sys/onoff.h

```diff
@@ -0,0 +1,611 @@
+ * has completed.
+ *
+ * Functions of this type are passed to service-specific transition
+ * functions to be used to report the completion of the operation.
+ * The functions may be invoked from any context.
+ *
+ * @param srv the service for which transition was requested.
+ *
+ * @param res the result of the transition.  This shall be
+ * non-negative on success, or a negative error code.  If an error is
+ * indicated the service shall enter an error state.
+ */
+typedef void (*onoff_service_notify_fn)(struct onoff_service *srv,
+					int res);
+
+/**
+ * @brief Signature used by service implementations to effect a
+ * transition.
+ *
+ * Service definitions use two function pointers of this type to be
+ * notified that a transition is required, and a third optional one to
+ * reset service state.
+ *
+ * The start function will be called only from the off state.
+ *
+ * The stop function will be called only from the on state.
+ *
+ * The reset function may be called only when
+ * `onoff_service_has_error()` returns true.
+ *
+ * Functions will be invoked from interrupt context only if the
+ * corresponding isr-capable flag is specified in the service
+ * configuration.
+ *
+ * @param srv the service for which transition was requested.
+ *
+ * @param notify the function to be invoked when the transition has
+ * completed.  The callee shall capture this parameter to notify on
+ * completion of asynchronous transitions.  If the transition is not
+ * asynchronous, notify shall be invoked before the transition
+ * function returns.
+ */
+typedef void (*onoff_service_transition_fn)(struct onoff_service *srv,
+					    onoff_service_notify_fn notify);
+
+/**
+ * @brief State associated with an on-off service.
+ *
+ * No fields in this structure are intended for use by service
+ * providers or clients.  The state is to be initialized once, using
+ * onoff_service_init(), when the service provider is initialized.
+ * In case of error it may be reset through the
+ * onoff_service_reset() API.
+ */
+struct onoff_service {
+	/* List of clients waiting for completion of reset or
+	 * transition to on.
+	 */
+	sys_slist_t clients;
+
+	/* Function to invoke to transition the service to on. */
+	onoff_service_transition_fn start;
+
+	/* Function to invoke to transition the service to off. */
+	onoff_service_transition_fn stop;
+
+	/* Function to force the service state to reset, where
+	 * supported.
+	 */
+	onoff_service_transition_fn reset;
+
+	/* Mutex protection for flags, clients, releaser, and refs. */
+	struct k_spinlock lock;
+
+	/* Client to be informed when transition to off completes. */
+	struct onoff_client *releaser;
+
+	/* Flags identifying the service state. */
+	u16_t flags;
+
+	/* Number of active clients for the service. */
+	u16_t refs;
+};
+
+/**
+ * @brief Compile-time initialization of an on-off service.
+ */
+#define ONOFF_SERVICE_INITIALIZER(_start, _stop, _reset, _flags) { \
```

Comment:
The intent here was to parallel what's done with `k_sem_init()` and `Z_SEM_INITIALIZER()` so people can implicitly initialize the service, if they know what they're doing.  I'll make this internal because explaining when and why you can and cannot do that is too complex.